### PR TITLE
Add a firstrest function

### PR DESCRIPTION
**Pull request by rofinn**
I often want to get the first element and an iterator of all remaining elements... figured other folks might also want to do this.

### CODE REVIEW

**Comment by oxinabox** [comment]
why not `head_and_tail`?

**Comment by rofinn** [comment]
> why not head_and_tail?

I figured this fit better with the existing naming convention as the behaviour matches calling `first(xs)` and `rest(xs, s)` which are existing functions.

File: src/IterTools.jl [hunk 1]
```diff
@@ -84,6 +85,27 @@ macro ifsomething(ex)
...
+(1, Base.Iterators.Rest{UnitRange{Int64},Int64}(1:3, 1))
+
+julia> collect(r)
+2-element Array{Int64,1}:
+ 2
+ 3
+```
+"""
+function firstrest(xs)
+    f, s = iterate(xs)
```
Comment (iamed2) [inline src/IterTools.jl]: If `iterate` returns `nothing`, this function should throw the same error as `first` does

File: test/runtests.jl [hunk 2]
```diff
@@ -10,6 +10,13 @@ include("testing_macros.jl")
 
 @testset "IterTools" begin
 @testset "iterators" begin
+    @testset "firstrest" begin
```
Comment (iamed2) [inline test/runtests.jl]: Should also test both empty and 1-item collections, as well as an array and generator (okay to just loop the same tests though).

File: src/IterTools.jl (hunk 1 above)
Comment (rofinn) [reply src/IterTools.jl]: Hmmm, that's different for each collection type though. I'm tempted to just throw an argument error for empty collections.

Comment (iamed2) [reply src/IterTools.jl]: Best to throw the one that generators throw I think (`ArgumentError: collection must be non-empty`)

Comment (rofinn) [reply src/IterTools.jl]: Alright, we're just throwing an argument error when empty and all other cases seem to work.

File: src/IterTools.jl [hunk 3]
```diff
@@ -101,6 +101,7 @@ julia> collect(r)
 ```
 """
 function firstrest(xs)
+    isempty(xs) && throw(ArgumentError("collection must be non-empty"))
```
Comment (iamed2) [inline src/IterTools.jl]: No we should just check if `iterate` returns `nothing`

Comment (rofinn) [reply src/IterTools.jl]: Why? That seems uglier.

Comment (iamed2) [reply src/IterTools.jl]: That's how the iteration interface works. Not everything implements `isempty` because not everything knows if it is empty until you iterate.

Comment (rofinn) [reply src/IterTools.jl]: Ahh, okay. That wasn't clear from the `isempty` docs.

Comment (iamed2) [reply src/IterTools.jl]: The only thing you can assume about an iterator is that it has `iterate`. And also the functions with default definitions: https://docs.julialang.org/en/latest/manual/interfaces/#man-interface-iteration-1

Comment (rofinn) [reply src/IterTools.jl]: Updated

Comment (rofinn) [reply src/IterTools.jl]: Updated

File: test/runtests.jl [hunk 4]
```diff
@@ -10,6 +10,36 @@ include("testing_macros.jl")
...
+        # We ranges/generators have different rest states vs array/tuples
```
Comment (iamed2) [inline test/runtests.jl]:
```suggestion
        # Ranges/generators have different rest states vs array/tuples
```

**Comment by sostock** [comment]
I’m kind of confused by this … the function is basically identical to `Base.Iterators.peel`, the only difference is that it throws a different error for an empty argument.

**Comment by iamed2** [comment]
Ah, yep, you're right.

We didn't find it because it's not documented.

**Comment by rofinn** [comment]
`peel` also seems like a weird name for it to me.
